File: restring/src/main/java/com/ice/restring/BottomNavigationViewTransformer.java
package com.ice.restring;

import android.content.res.Resources;
import android.content.res.XmlResourceParser;
import android.support.design.widget.BottomNavigationView;
import android.util.AttributeSet;
import android.util.Pair;
import android.util.Xml;
import android.view.View;

import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserException;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

/**
 * A transformer which transforms BottomNavigationView: it transforms the texts coming from the menu.
 */
class BottomNavigationViewTransformer implements ViewTransformerManager.Transformer {

    private static final String ATTRIBUTE_MENU = "menu";
    private static final String ATTRIBUTE_APP_MENU = "app:menu";
    private static final String ATTRIBUTE_ID = "id";
    private static final String ATTRIBUTE_ANDROID_ID = "android:id";
    private static final String ATTRIBUTE_TITLE = "title";
    private static final String ATTRIBUTE_ANDROID_TITLE = "android:title";
    private static final String ATTRIBUTE_TITLE_CONDENSED = "titleCondensed";
    private static final String ATTRIBUTE_ANDROID_TITLE_CONDENSED = "android:titleCondensed";
    private static final String XML_MENU = "menu";
    private static final String XML_ITEM = "item";

    @Override
    public Class<? extends View> getViewType() {
        return BottomNavigationView.class;
    }

    @Override
    public View transform(View view, AttributeSet attrs) {
        if (view == null || !getViewType().isInstance(view)) {
            return view;
        }
        Resources resources = view.getContext().getResources();
        BottomNavigationView bottomNavigationView = (BottomNavigationView) view;

        for (int index = 0; index < attrs.getAttributeCount(); index++) {
            String attributeName = attrs.getAttributeName(index);
            switch (attributeName) {
                case ATTRIBUTE_APP_MENU:
                case ATTRIBUTE_MENU: {
                    String value = attrs.getAttributeValue(index);
                    if (value == null || !value.startsWith("@")) break;

                    int resId = attrs.getAttributeResourceValue(index, 0);
                    Map<Integer, MenuItemStrings> itemStrings = getMenuItemsStrings(resources, resId);

                    for (Map.Entry<Integer, MenuItemStrings> entry : itemStrings.entrySet()) {

                        if (entry.getValue().title != 0) {
                            bottomNavigationView.getMenu().findItem(entry.getKey()).setTitle(
                                    resources.getString(entry.getValue().title)
                            );
                        }
                        if (entry.getValue().titleCondensed != 0) {
                            bottomNavigationView.getMenu().findItem(entry.getKey()).setTitleCondensed(
                                    resources.getString(entry.getValue().titleCondensed)
                            );
                        }
                    }

                    break;
                }
            }
        }

        return view;
    }

    private Map<Integer, MenuItemStrings> getMenuItemsStrings(Resources resources, int resId) {
        XmlResourceParser parser = resources.getLayout(resId);
        AttributeSet attrs = Xml.asAttributeSet(parser);
        try {
            return parseMenu(parser, attrs);
        } catch (XmlPullParserException | IOException e) {
            e.printStackTrace();
            return new HashMap<>();
        }
    }

    private Map<Integer, MenuItemStrings> parseMenu(XmlPullParser parser, AttributeSet attrs)
            throws XmlPullParserException, IOException {

        Map<Integer, MenuItemStrings> menuItems = new HashMap<>();
        int eventType = parser.getEventType();
        String tagName;

        // This loop will skip to the menu start tag
        do {
            if (eventType == XmlPullParser.START_TAG) {
                tagName = parser.getName();
                if (tagName.equals(XML_MENU)) {
                    eventType = parser.next();
                    break;
                }

                throw new RuntimeException("Expecting menu, got " + tagName);
            }
            eventType = parser.next();
        } while (eventType != XmlPullParser.END_DOCUMENT);

        boolean reachedEndOfMenu = false;
        int menuLevel = 0;
        while (!reachedEndOfMenu) {
            switch (eventType) {
                case XmlPullParser.START_TAG:
                    tagName = parser.getName();
                    if (tagName.equals(XML_ITEM)) {
                        Pair<Integer, MenuItemStrings> item = parseMenuItem(attrs);
                        if (item != null) {
                            menuItems.put(item.first, item.second);
                        }
                    } else if (tagName.equals(XML_MENU)) {
                        menuLevel++;
                    }
                    break;

                case XmlPullParser.END_TAG:
                    tagName = parser.getName();
                    if (tagName.equals(XML_MENU)) {
                        menuLevel--;
                        if (menuLevel <= 0) {
                            reachedEndOfMenu = true;
                        }
                    }
                    break;

                case XmlPullParser.END_DOCUMENT:
                    reachedEndOfMenu = true;
            }

            eventType = parser.next();
        }
        return menuItems;
    }

    private Pair<Integer, MenuItemStrings> parseMenuItem(AttributeSet attrs) {
        int menuId = 0;
        MenuItemStrings menuItemStrings = null;
        int attributeCount = attrs.getAttributeCount();
        for (int index = 0; index < attributeCount; index++) {
            switch (attrs.getAttributeName(index)) {
                case ATTRIBUTE_ANDROID_ID:
                case ATTRIBUTE_ID: {
                    menuId = attrs.getAttributeResourceValue(index, 0);
                    break;
                }
                case ATTRIBUTE_ANDROID_TITLE:
                case ATTRIBUTE_TITLE: {
                    String value = attrs.getAttributeValue(index);
                    if (value == null || !value.startsWith("@")) break;
                    if (menuItemStrings == null) {
                        menuItemStrings = new MenuItemStrings();
                    }
                    menuItemStrings.title = attrs.getAttributeResourceValue(index, 0);
                    break;
                }
                case ATTRIBUTE_ANDROID_TITLE_CONDENSED:
                case ATTRIBUTE_TITLE_CONDENSED: {
                    String value = attrs.getAttributeValue(index);
                    if (value == null || !value.startsWith("@")) break;
                    if (menuItemStrings == null) {
                        menuItemStrings = new MenuItemStrings();
                    }
                    menuItemStrings.titleCondensed = attrs.getAttributeResourceValue(index, 0);
                    break;
                }
            }
        }
        return (menuId != 0 && menuItemStrings != null)
                ? new Pair<>(menuId, menuItemStrings)
                : null;
    }

    private static class MenuItemStrings {
        public int title;
        public int titleCondensed;
    }
}

File: restring/src/main/java/com/b3nedikt/restring/BottomNavigationViewTransformer.kt
package com.b3nedikt.restring

import android.content.res.Resources
import android.util.AttributeSet
import android.util.Pair
import android.util.Xml
import android.view.View
import com.google.android.material.bottomnavigation.BottomNavigationView
import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserException
import java.io.IOException
import java.util.*

/**
 * A transformer which transforms BottomNavigationView: it transforms the texts coming from the menu.
 */
internal class BottomNavigationViewTransformer : ViewTransformerManager.Transformer {

    override val viewType = BottomNavigationView::class.java

    override fun transform(view: View, attrs: AttributeSet): View {
        if (!viewType.isInstance(view)) {
            return view
        }
        val resources = view.context.resources
        val bottomNavigationView = view as BottomNavigationView

        for (index in 0 until attrs.attributeCount) {
            val attributeName = attrs.getAttributeName(index)
            if (attributeName == ATTRIBUTE_APP_MENU || attributeName == ATTRIBUTE_MENU) {
                val value = attrs.getAttributeValue(index)
                if (value == null || !value.startsWith("@")) break

                val resId = attrs.getAttributeResourceValue(index, 0)
                val itemStrings = getMenuItemsStrings(resources, resId)

                for ((key, value1) in itemStrings) {

                    if (value1.title != 0) {
                        bottomNavigationView.menu.findItem(key).title = resources.getString(value1.title)
                    }
                    if (value1.titleCondensed != 0) {
                        bottomNavigationView.menu.findItem(key).titleCondensed = resources.getString(value1.titleCondensed)
                    }
                }
            }
        }

        return view
    }

    private fun getMenuItemsStrings(resources: Resources, resId: Int): Map<Int, MenuItemStrings> {
        val parser = resources.getLayout(resId)
        val attrs = Xml.asAttributeSet(parser)
        return try {
            parseMenu(parser, attrs)
        } catch (e: XmlPullParserException) {
            e.printStackTrace()
            HashMap()
        } catch (e: IOException) {
            e.printStackTrace()
            HashMap()
        }

    }

    @Throws(XmlPullParserException::class, IOException::class)
    private fun parseMenu(parser: XmlPullParser, attrs: AttributeSet): Map<Int, MenuItemStrings> {

        val menuItems = mutableMapOf<Int, MenuItemStrings>()
        var eventType = parser.eventType
        var tagName: String

        // This loop will skip to the menu start tag
        do {
            if (eventType == XmlPullParser.START_TAG) {
                tagName = parser.name
                if (tagName == XML_MENU) {
                    eventType = parser.next()
                    break
                }

                throw RuntimeException("Expecting menu, got $tagName")
            }
            eventType = parser.next()
        } while (eventType != XmlPullParser.END_DOCUMENT)

        var reachedEndOfMenu = false
        var menuLevel = 0
        while (!reachedEndOfMenu) {
            when (eventType) {
                XmlPullParser.START_TAG -> {
                    tagName = parser.name
                    if (tagName == XML_ITEM) {
                        val item = parseMenuItem(attrs)
                        if (item != null) {
                            menuItems[item.first] = item.second
                        }
                    } else if (tagName == XML_MENU) {
                        menuLevel++
                    }
                }

                XmlPullParser.END_TAG -> {
                    tagName = parser.name
                    if (tagName == XML_MENU) {
                        menuLevel--
                        if (menuLevel <= 0) {
                            reachedEndOfMenu = true
                        }
                    }
                }

                XmlPullParser.END_DOCUMENT -> reachedEndOfMenu = true
            }

            eventType = parser.next()
        }
        return menuItems
    }

    private fun parseMenuItem(attrs: AttributeSet): Pair<Int, MenuItemStrings>? {
        var menuId = 0
        var menuItemStrings: MenuItemStrings? = null
        val attributeCount = attrs.attributeCount
        for (index in 0 until attributeCount) {
            if (attrs.getAttributeName(index) == ATTRIBUTE_ANDROID_ID
                    || attrs.getAttributeName(index) == ATTRIBUTE_ID) {
                menuId = attrs.getAttributeResourceValue(index, 0)
            }
            else if (attrs.getAttributeName(index) == ATTRIBUTE_ANDROID_TITLE
                    || attrs.getAttributeName(index) == ATTRIBUTE_TITLE) {
                val value = attrs.getAttributeValue(index)
                if (value == null || !value.startsWith("@")) break
                if (menuItemStrings == null) {
                    menuItemStrings = MenuItemStrings()
                }
                menuItemStrings.title = attrs.getAttributeResourceValue(index, 0)
            }
            else if (attrs.getAttributeName(index) == ATTRIBUTE_ANDROID_TITLE_CONDENSED
                    || attrs.getAttributeName(index) == ATTRIBUTE_TITLE_CONDENSED) {
                val value = attrs.getAttributeValue(index)
                if (value == null || !value.startsWith("@")) break
                if (menuItemStrings == null) {
                    menuItemStrings = MenuItemStrings()
                }
                menuItemStrings.titleCondensed = attrs.getAttributeResourceValue(index, 0)
            }
        }
        return if (menuId != 0 && menuItemStrings != null)
            Pair(menuId, menuItemStrings)
        else
            null
    }

    private class MenuItemStrings {
        var title: Int = 0
        var titleCondensed: Int = 0
    }

    private companion object {
        private const val ATTRIBUTE_MENU = "menu"
        private const val ATTRIBUTE_APP_MENU = "app:menu"
        private const val ATTRIBUTE_ID = "id"
        private const val ATTRIBUTE_ANDROID_ID = "android:id"
        private const val ATTRIBUTE_TITLE = "title"
        private const val ATTRIBUTE_ANDROID_TITLE = "android:title"
        private const val ATTRIBUTE_TITLE_CONDENSED = "titleCondensed"
        private const val ATTRIBUTE_ANDROID_TITLE_CONDENSED = "android:titleCondensed"
        private const val XML_MENU = "menu"
        private const val XML_ITEM = "item"
    }
}
